x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java

```diff
@@ -533,22 +542,27 @@ public void authenticate(final AuthenticationToken authenticationToken, final Ac
                 JwtValidateUtil.validateExpiredTime(jwt, now, this.allowedClockSkew.seconds());
 
                 // Validate signature last, so expensive JWK reloads only happen after all other checks passed.
-                final boolean reloadBefore = this.contentAndFilteredJwksAlgsHmac.isEmpty() && this.contentAndFilteredJwksAlgsPkc.isEmpty();
-                final ContentAndFilteredJwksAlgs contentAndFilteredJwksAlgs = isJwtSignatureAlgHmac
-                    ? this.contentAndFilteredJwksAlgsHmac
-                    : this.contentAndFilteredJwksAlgsPkc;
                 try {
-                    if ((reloadBefore) && (this.reloadJwksHelper(listener, tokenPrincipal, isJwtSignatureAlgHmac) == false)) {
-                        return; // Reload required before. Reload failed, or succeeded but no changes. Helper populates listener response.
+                    if (this.contentAndFilteredJwksAlgsHmac.isEmpty() && this.contentAndFilteredJwksAlgsPkc.isEmpty()) {
+                        LOGGER.debug("Realm [" + super.name() + "] has no JWKs and Algs to verify JWT for token=[" + tokenPrincipal + "].");
+                        if (this.reloadJwksHelper(listener, tokenPrincipal, isJwtSignatureAlgHmac) == false) {
+                            return; // Try reload before. Reload failed, or succeeded but no changes. Helper handles listener and logs.
+                        }
                     }
-                    // Throws AllVerifiesFailedException (some JWKs after filtering, all failed) or Exception (no JWKs after filtering)
+                    // Throws Exception (no JWKs after filtering) or AllVerifiesFailedException (all available JWKs failed)
+                    final ContentAndFilteredJwksAlgs contentAndFilteredJwksAlgs = isJwtSignatureAlgHmac
```

Comment:
Contents were replaced a few lines above by reloadJwksHelper, so revalidate needs to use the new, filtered alg+jwk references.